api/src/main/java/net/kyori/adventure/text/Component.java

```diff
@@ -143,20 +145,86 @@
    * @param components the components
    * @return a text component
    * @since 4.0.0
+   * @deprecated for removal since 4.8.0, use {@link #join(JoinConfiguration, Iterable)} with a {@link JoinConfiguration} instead.
    */
+  @Deprecated
   @Contract(value = "_, _ -> new", pure = true)
   static @NonNull TextComponent join(final @NonNull ComponentLike separator, final Iterable<? extends ComponentLike> components) {
+    return join(JoinConfiguration.separator(separator), components);
+  }
+
+  /**
+   * Joins {@code components} using the configuration in {@code config}.
+   *
+   * @param config the join configuration
+   * @param components the components
+   * @return a text component
+   * @see JoinConfiguration
+   * @since 4.8.0
+   */
+  @Contract(value = "_, _ -> new", pure = true)
+  static @NonNull TextComponent join(final @NonNull JoinConfiguration config, final @NonNull ComponentLike@NonNull... components) {
+    return join(config, Arrays.asList(components));
+  }
+
+  /**
+   * Joins {@code components} using the configuration in {@code config}.
+   *
+   * @param config the join configuration
+   * @param components the components
+   * @return a text component
+   * @see JoinConfiguration
+   * @since 4.8.0
+   */
+  @Contract(value = "_, _ -> new", pure = true)
+  static @NonNull TextComponent join(final @NonNull JoinConfiguration config, final @NonNull Iterable<? extends ComponentLike> components) {
```

Comment:
Would this API be more idiomatic as a builder, or a modifiable (albeit immutable) `ComponentJoiner` object?

```java
  private static final ComponentJoiner PLUGIN_JOINER = Component.joiner(Component.text(", "))
    .prefix(prefix)
    .suffix(suffix);

  // ...

  public static void sendPluginList(Audience target) {
    target.sendMessage(PLUGIN_JOINER.join(getPluginNameComponents());
  }
```

Reply:
There is precedent here: see Guava `Joiner` and `Splitter` which uses the pattern I've described here.

Reply:
This PR is matching how we do this in the rest of the Adventure API, however.

Reply:
I'm still not a big fan of it. I guess it's something that can be revisited in Adventure 5, because I still believe this pattern feels unnatural to write.

Reply:
I do agree that a shorthand method to join directly from a join configuration would be nice, an earlier version of the PR did include one. Although I do understand that the syntax doesn't differ too much whether you're doing `config.join(components)` or `join(config, components)` and the latter is certainly more in keeping with the rest of the API, such as the text replacement system.